src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorHtmlPublishDiagnosticsInterceptor.cs

```diff
@@ -0,0 +1,195 @@
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            var diagnosticParams = token.ToObject<VSPublishDiagnosticParams>();
+
+            if (diagnosticParams is null)
+            {
+                throw new ArgumentException("Conversion of token failed.");
+            }
+
+            // We only support interception of Virtual HTML Files
+            if (!RazorLSPConventions.IsVirtualHtmlFile(diagnosticParams.Uri))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            var razorDocumentUri = RazorLSPConventions.GetRazorDocumentUri(diagnosticParams.Uri);
+
+            if (!_documentManager.TryGetDocument(razorDocumentUri, out var razorDocumentSnapshot))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            if (!razorDocumentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlDocumentSnapshot))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            // Ensure we're working with the virtual document specified in the diagnostic params
+            if (!htmlDocumentSnapshot.Uri.Equals(diagnosticParams.Uri))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            // Note; this is an `interceptor` & not a handler, hence 
+            // it's possible another interceptor mutates this request 
+            // later in the toolchain. Such an interceptor would likely 
+            // expect a `__virtual.html` suffix instead of `.razor`.
+            diagnosticParams.Uri = razorDocumentUri;
+
+            // Return early if there aren't any diagnostics to process
+            if (diagnosticParams.Diagnostics?.Any() != true)
+            {
+                return CreateResponse(diagnosticParams);
+            }
+
+            var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync(
+                razorDocumentSnapshot.Version,
+                htmlDocumentSnapshot,
+                cancellationToken).ConfigureAwait(false);
+            if (!synchronized)
+            {
+                // Could not synchronize, we'll clear out the diagnostics as we don't
+                // know whether or not they're still valid in the current state of the
+                // document.
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var unmappedDiagnostics = diagnosticParams.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d));
+            if (!filteredDiagnostics.Any())
+            {
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var rangesToMap = filteredDiagnostics.Select(r => r.Range).ToArray();
+            var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync(
```

Comment:
Don't we want to have the filter/remap logic in the Razor language server? That way all of this mapping/filtering can exist in one place.

Reply:
Yep 💯 , will handle this and the same thing for C# Pull diagnostics via https://github.com/dotnet/aspnetcore/issues/27563. Reason I'm not handling it all in this PR is because this'll require moving a bunch of things around and bloat this PR (admittedly the RemappingResponse changes probably belongs in it's own PR anyways). 

Going to hold off in the test cases as well for now given they'll need to be re-written following the refactor, hope that's alright (changes should be isolated to this file only anyways).